File: retirodelrocioapp/lib/features/kitchen/domain/kitchen_notification.dart
import 'package:flutter/foundation.dart';

/// A new-order alert on the Kitchen Tablet's notification feed
/// (`GET /kitchen/notifications`).
@immutable
class KitchenNotification {
  const KitchenNotification({
    required this.id,
    required this.category,
    required this.title,
    required this.message,
    required this.diningOrderId,
    required this.orderCode,
    required this.createdLabel,
    required this.read,
  });

  final int id;
  final String category;
  final String title;
  final String message;
  final int? diningOrderId;
  final String? orderCode;
  final String? createdLabel;
  final bool read;

  factory KitchenNotification.fromJson(Map<String, dynamic> json) =>
      KitchenNotification(
        id: (json['id'] as num?)?.toInt() ?? 0,
        category: json['category'] as String? ?? '',
        title: json['title'] as String? ?? '',
        message: json['message'] as String? ?? '',
        diningOrderId: (json['dining_order_id'] as num?)?.toInt(),
        orderCode: json['order_code'] as String?,
        createdLabel: json['created_label'] as String?,
        read: json['read'] as bool? ?? false,
      );
}

File: retiro-del-rocio/app/Models/KitchenNotification.php
<?php

namespace App\Models;

use App\Events\KitchenNotificationSent;
use Illuminate\Database\Eloquent\Model;
use Throwable;

/**
 * A food-order-relevant update pushed to the kitchen — feeds the Kitchen
 * Tablet's new-order alert popup and notification feed. Kitchen is one
 * shared station, so the feed (and its read state) is shared across whoever
 * is signed in, same as bar's.
 *
 * Trigger: a new dining order containing a food item lands — see
 * {@see DiningOrder}'s `booted()` hook.
 */
class KitchenNotification extends Model
{
    protected $fillable = [
        'dining_order_id', 'category', 'title', 'message', 'read_at',
    ];

    protected $casts = [
        'read_at' => 'datetime',
    ];

    public function diningOrder()
    {
        return $this->belongsTo(DiningOrder::class);
    }

    /**
     * Create a kitchen notification and push a live signal over the
     * hotel-wide `kitchen` channel. Broadcasting must never break the
     * calling flow — if the socket server is down the notification is
     * still recorded, and the tablet's own poll picks it up.
     */
    public static function notify(string $category, string $title, string $message, ?DiningOrder $order = null): self
    {
        $notification = static::create([
            'dining_order_id' => $order?->id,
            'category' => $category,
            'title' => $title,
            'message' => $message,
        ]);

        try {
            broadcast(new KitchenNotificationSent);
        } catch (Throwable $e) {
            report($e);
        }

        return $notification;
    }

    public function toKitchenArray(): array
    {
        $order = $this->relationLoaded('diningOrder') ? $this->diningOrder : $this->diningOrder()->first();

        return [
            'id' => $this->id,
            'category' => $this->category,
            'title' => $this->title,
            'message' => $this->message,
            'dining_order_id' => $this->dining_order_id,
            'order_code' => $order?->orderCode(),
            'created_at' => $this->created_at?->toIso8601String(),
            'created_label' => optional($this->created_at)->diffForHumans(),
            'read' => $this->read_at !== null,
        ];
    }
}
